### PR TITLE
Replaced usage of ObjectManager with Constructor DI of Factory class

### PR DESCRIPTION
**Pull request by drpayyne**
### Description
Fixed the improper usage of ObjectManager to get instance of Factory class in the middle of a method, instead of using Constructor Dependency Injection, in `Magento_Catalog/Controller/Adminhtml/Category/Add`.
- Removed usage of ObjectManager
- Injected `\Magento\Framework\View\Result\PageFactory` through Constructor DI
- Fixed return type annotation in appropriate PHPDoc Blocks for method and constructor

### Fixed Issues
1. magento/magento2#24646: Replacing the usage of ObjectManager in Magento_Catalog Controllers

### Manual testing scenarios
No practical testing, but
1. Code changes in `Magento_Catalog/Controller/Adminhtml/Category/Add` as described above

### Questions or comments
Bug fix, as usage of ObjectManager directly in code is prohibited.

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/magento2?pullRequest=24661) <br/>All committers have signed the CLA.

File: app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php [hunk 1]
```diff
@@ -61,9 +72,8 @@ public function execute()
             $category->addData($categoryData);
         }
 
-        $resultPageFactory = $this->_objectManager->get(\Magento\Framework\View\Result\PageFactory::class);
-        /** @var \Magento\Backend\Model\View\Result\Page $resultPage */
-        $resultPage = $resultPageFactory->create();
+        /** @var Page $resultPage */
+        $resultPage = $this->resultPageFactory->create();
```
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php]: Hi @drpayyne,
AFAIK there should be available factory for all types of results - `$this->resultFactory`. Is so - we should use it instead of introducing new dependencies to constructor.
Could you check that?

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

File: app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php (hunk 1 above)
Comment (joshua-bn) [reply app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php]: https://github.com/magento/magento2/blob/2.3-develop/lib/internal/Magento/Framework/Controller/ResultFactory.php

File: app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php [hunk 2]
```diff
@@ -22,32 +25,40 @@ class Add extends \Magento\Catalog\Controller\Adminhtml\Category implements Http
...
         parent::__construct($context);
         $this->resultForwardFactory = $resultForwardFactory;
+        $this->resultPageFactory = $resultPageFactory;
     }
 
     /**
      * Add new category form
      *
-     * @return \Magento\Backend\Model\View\Result\Forward
+     * @return Page | Redirect | Json
```
Comment (joshua-bn) [inline app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php]: This should return the ResultInterface: https://github.com/magento/magento2/blob/2.3-develop/lib/internal/Magento/Framework/Controller/ResultInterface.php which is what is returned by the ResultFactory https://github.com/magento/magento2/blob/2.3-develop/lib/internal/Magento/Framework/Controller/ResultFactory.php#L81

Comment (joshuaadickerson) [inline app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php]: This should return the ResultInterface: https://github.com/magento/magento2/blob/2.3-develop/lib/internal/Magento/Framework/Controller/ResultInterface.php which is what is returned by the ResultFactory https://github.com/magento/magento2/blob/2.3-develop/lib/internal/Magento/Framework/Controller/ResultFactory.php#L81

**Review by joshuaadickerson — CHANGES_REQUESTED** [review]
Use return types too :)

File: app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php (hunk 2 above)
Comment (drpayyne) [reply app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php]: Made the suggested changes. Thanks!

File: app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php (hunk 1 above)
Comment (drpayyne) [reply app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php]: Fixed the code according to your requested changes, @ihor-sviziev. Thanks!

**Review by joshuaadickerson — APPROVED** [review]

File: app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php [hunk 3]
```diff
@@ -6,49 +6,55 @@
...
 {
     /**
-     * Forward factory for result
+     * Common Factory to create any Result object
      *
-     * @var \Magento\Backend\Model\View\Result\ForwardFactory
+     * @var ResultFactory
      */
-    protected $resultForwardFactory;
+    protected $resultFactory;
```
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php]: Hi,
`$this->resultFactory` already defined as protected property in `\Magento\Framework\App\Action\AbstractAction`, as this class extends it - it's already there, we don't need to add it again.

File: app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php [hunk 4]
```diff
@@ -6,49 +6,55 @@
...
-     * @param \Magento\Backend\App\Action\Context $context
-     * @param \Magento\Backend\Model\View\Result\ForwardFactory $resultForwardFactory
+     * @param Context $context
+     * @param ResultFactory $resultFactory
      */
     public function __construct(
-        \Magento\Backend\App\Action\Context $context,
-        \Magento\Backend\Model\View\Result\ForwardFactory $resultForwardFactory
+        Context $context,
+        ResultFactory $resultFactory
```
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php]: Please revert this change, it's not backward compatible and it's actually not needed

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @drpayyne,
Please review my comments, once you'll finish - please squash all your changes to single commit & force push them.

Thank you!

File: app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php (hunk 4 above)
Comment (drpayyne) [reply app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php]: Yes, I'll do that. But that'll make `$resultForwardFactory` unused code, right?

File: app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php (hunk 3 above)
Comment (drpayyne) [reply app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php]: Done, thank you.

File: app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php (hunk 4 above)
Comment (drpayyne) [reply app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php]: Done, thank you.

**Review by ihor-sviziev — APPROVED** [review]

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-5909](https://jira.corp.magento.com/browse/ENGCOM-5909) has been created to process this Pull Request

**Comment by magento-engcom-team** [comment]
@drpayyne thank you for contributing. Please accept Community Contributors team invitation [here](https://github.com/orgs/magento/invitation) to gain extended permissions for this repository.

File: app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php [hunk 5]
```diff
@@ -6,31 +6,38 @@
...
 
-use Magento\Framework\App\Action\HttpGetActionInterface as HttpGetActionInterface;
+use Magento\Backend\App\Action\Context;
+use Magento\Backend\Model\View\Result\Page;
+use Magento\Framework\Controller\ResultFactory;
+use Magento\Backend\Model\View\Result\Redirect;
+use Magento\Framework\Controller\ResultInterface;
+use Magento\Catalog\Controller\Adminhtml\Category;
+use Magento\Backend\Model\View\Result\ForwardFactory;
+use Magento\Framework\App\Action\HttpGetActionInterface;
```
Comment (orlangur) [inline app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php:16]: No need to revert, but please avoid importing FQCN mixed with other changes.

File: app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php [hunk 6]
```diff
@@ -39,16 +46,16 @@ public function __construct(
     /**
      * Add new category form
      *
-     * @return \Magento\Backend\Model\View\Result\Forward
+     * @return ResultInterface
      */
-    public function execute()
+    public function execute(): ResultInterface
```
Comment (orlangur) [inline app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php]: I don't think it is backward compatible.

File: app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php [hunk 7]
```diff
@@ -39,16 +46,16 @@ public function __construct(
...
     {
-        $parentId = (int)$this->getRequest()->getParam('parent');
+        $parentId = (int) $this->getRequest()->getParam('parent');
```
Comment (orlangur) [inline app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php]: Revert.

File: app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php [hunk 8]
```diff
@@ -61,9 +68,8 @@ public function execute()
             $category->addData($categoryData);
         }
 
-        $resultPageFactory = $this->_objectManager->get(\Magento\Framework\View\Result\PageFactory::class);
-        /** @var \Magento\Backend\Model\View\Result\Page $resultPage */
-        $resultPage = $resultPageFactory->create();
+        /** @var Page $resultPage */
+        $resultPage = $this->resultFactory->create(ResultFactory::TYPE_PAGE);
```
Comment (orlangur) [inline app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php:75]: Will check if such change is needed together with other PR.

**Review by orlangur — CHANGES_REQUESTED** [review]

File: app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php (hunk 5 above)
Comment (ihor-sviziev) [reply app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php:16]: > No need to revert, but please avoid importing FQCN mixed with other changes.

Hi @orlangur, Why we should avoid importing FQCN mixed with other changes?
It's just making code simplier

File: app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php (hunk 6 above)
Comment (ihor-sviziev) [reply app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php]: > I don't think it is backward compatible.

There were no complains about backward incompatible change from static tests, usually we do have

File: app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php (hunk 7 above)
Comment (ihor-sviziev) [reply app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php]: > Revert.

Why should we revert it? There were no complains about it from static tests and actually seems ok to me.

File: app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php (hunk 8 above)
Comment (ihor-sviziev) [reply app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php:75]: It's needed because `$this->resultFactory` we have in all admin controllers, it comes from `Magento\Backend\App\Action\Context` in constructor. It's better to have one dependency, that already was there, rather than introducing new and new factory dependencies for all types of results.

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-5909](https://jira.corp.magento.com/browse/ENGCOM-5909) has been created to process this Pull Request

File: app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php (hunk 6 above)
Comment (orlangur) [reply app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php]: Hm. @sidolov, remember you mentioned in some of PRs adding return type even for controller action is not compatible? Do we have all BC violations reported by automated tests now?

File: app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php (hunk 7 above)
Comment (orlangur) [reply app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php]: Because it must not be touched, common formatting for Magento code is without space (not enforced by any test yet though).

File: app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php [hunk 9]
```diff
@@ -6,31 +6,38 @@
...
+     * @param ForwardFactory $resultForwardFactory
      */
     public function __construct(
-        \Magento\Backend\App\Action\Context $context,
-        \Magento\Backend\Model\View\Result\ForwardFactory $resultForwardFactory
+        Context $context,
+        ForwardFactory $resultForwardFactory
     ) {
         parent::__construct($context);
         $this->resultForwardFactory = $resultForwardFactory;
```
Comment (orlangur) [inline app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php:46]: Looks like this field can be marked as deprecated.

File: app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php [hunk 10]
```diff
@@ -39,16 +46,16 @@ public function __construct(
...
     {
-        $parentId = (int)$this->getRequest()->getParam('parent');
+        $parentId = (int) $this->getRequest()->getParam('parent');
 
         $category = $this->_initCategory(true);
         if (!$category || !$parentId || $category->getId()) {
-            /** @var \Magento\Backend\Model\View\Result\Redirect $resultRedirect */
-            $resultRedirect = $this->resultRedirectFactory->create();
+            /** @var Redirect $resultRedirect */
+            $resultRedirect = $this->resultFactory->create(ResultFactory::TYPE_REDIRECT);
```
Comment (orlangur) [inline app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php]: No reason to change this, please revert.

File: app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php (hunk 8 above)
Comment (orlangur) [reply app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php:75]: @ihor-sviziev I'm more concerned regarding `$this->resultRedirectFactory` not used.

I'm ok with this OM replacement.

**Review by orlangur — CHANGES_REQUESTED** [review]

File: app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php (hunk 9 above)
Comment (drpayyne) [reply app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php:46]: How do I deprecate a field? An example would help, thanks. @orlangur

File: app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php [hunk 11]
```diff
@@ -6,31 +6,38 @@
...
  * @package Magento\Catalog\Controller\Adminhtml\Category
  */
-class Add extends \Magento\Catalog\Controller\Adminhtml\Category implements HttpGetActionInterface
+class Add extends Category implements HttpGetActionInterface
 {
     /**
      * Forward factory for result
      *
-     * @var \Magento\Backend\Model\View\Result\ForwardFactory
+     * @var ForwardFactory
```
Comment (orlangur) [inline app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php:30]:
```suggestion
     * @deprecated
     * @var ForwardFactory
```

Comment (orlangur) [reply app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php:30]: Please squash it with your changes ofc

Comment (ihor-sviziev) [reply app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php:30]: Please add comment why it's deprecated.

Comment (orlangur) [reply app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php:30]: Just because it's not used? I'm not even sure it was ever used

Comment (ihor-sviziev) [reply app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php:30]: just wanted to add description why it was deprecated, like this
```suggestion
     * @deprecated not used anymore
     * @see $this->resultFactory->create()
     * @var ForwardFactory
```

Comment (drpayyne) [reply app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php:30]: Done, thank you.

File: app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php [hunk 12]
```diff
@@ -39,15 +47,15 @@ public function __construct(
...
      */
     public function execute()
     {
         $parentId = (int)$this->getRequest()->getParam('parent');
 
         $category = $this->_initCategory(true);
         if (!$category || !$parentId || $category->getId()) {
-            /** @var \Magento\Backend\Model\View\Result\Redirect $resultRedirect */
+            /** @var Redirect $resultRedirect */
             $resultRedirect = $this->resultRedirectFactory->create();
```
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php:61]: @orlangur I think it's better to replace "resultRedirectFactory", "resultForwardFactory", "resultPageFactory" to `$this->resultFactory->create(<type>);` in all places to have in all places similar code. Why you think it's bad idea?

File: app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php (hunk 5 above)
Comment (ihor-sviziev) [reply app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php:16]: @orlangur ping

File: app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php [hunk 13]
```diff
@@ -6,31 +6,39 @@
...
      */
     protected $resultForwardFactory;
 
     /**
      * Add category constructor
      *
-     * @param \Magento\Backend\App\Action\Context $context
-     * @param \Magento\Backend\Model\View\Result\ForwardFactory $resultForwardFactory
+     * @param Context $context
+     * @deprecated Unused Class: ForwardFactory
```
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php]: It should be not there, near property definition, phpdoc block for `resultForwardFactory`.
See https://github.com/magento/magento2/pull/24661/commits/9be9930236b3ecaa473f1e51e57a67ef95a7ee2b#diff-30a41b01a5dfb806442dd6e6b6775a15R28

File: app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php (hunk 9 above)
Comment (ihor-sviziev) [reply app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php:46]: > How do I deprecate a field? An example would help, thanks. @orlangur

See https://github.com/magento/magento2/pull/24661/commits/9be9930236b3ecaa473f1e51e57a67ef95a7ee2b#diff-30a41b01a5dfb806442dd6e6b6775a15R28

File: app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php (hunk 13 above)
Comment (drpayyne) [reply app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php]: Fixed, sorry about that @ihor-sviziev.

File: app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php (hunk 12 above)
Comment (ihor-sviziev) [reply app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php:61]: In order to highlight the issue scope more I created PR that has description. https://github.com/magento/magento2/pull/24711

Let's discuss it separately.

Comment (ihor-sviziev) [reply app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php:61]: For now let's keep it as is

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-5909](https://jira.corp.magento.com/browse/ENGCOM-5909) has been created to process this Pull Request

File: app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php (hunk 5 above)
Comment (orlangur) [reply app/code/Magento/Catalog/Controller/Adminhtml/Category/Add.php:16]: @ihor-sviziev it should be delivered as a separate efforts not mixed with other changes.

**Review by orlangur — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @orlangur, thank you for the review.
[ENGCOM-5909](https://jira.corp.magento.com/browse/ENGCOM-5909) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed
